### ISSUE TITLE
MDI support?

### ISSUE THREAD

**Issue by Shadowblitz16** [libui-ng/libui-ng#129]
I know MDI would be easy to implement on windows, but is it possible to get this to work on Mac and Linux without using some other implementation on other platforms?

It's basically just a window clipped and offset by the parent window, with the parent window having some sort of darker colored inset panel.

**Comment by cody271** [comment]
> I know MDI would be easy to implement on windows, but is it possible to get this to work on Mac and Linux without using some other implementation on other platforms?

MDI is not really possible cross platform natively no, that was a Windows only concept. Maybe with some hacks, can sorta kinda do it on Unix. Not possible at all on Mac AFAIK.

I can think of some other options though:

1) Use uiTab to organize the child "windows" inside your main window.

2) Self drawn controls with uiArea, for a hybrid mix of native and self drawn content.

You can also take a look at how wxWidgets attempts to emulate MDI as a fallback on non-Windows platforms for another idea on how to implement this.

Linking to your [original](https://github.com/andlabs/libui/issues/475) [issues](https://github.com/simplexidev/sdfx/issues/38) on the topic here as well.

**Comment by Shadowblitz16** [comment]
@cody271 I dislike how wxwidgets does it which uses tabs on unix.

Maybe drawing custom controls would work if you could find a way to render the title bar natively?
Idk if those methods are exposed on mac or not.

**Comment by cody271** [comment]
@Shadowblitz16 Are you able to share more details of your application? Based on previous discussion, this is a game editor yes?

For this use case, your best bet is to use a hybrid approach where your main window content is 100% self drawn in your own theme. But then the menu bars / dialogs / preferences / etc can still all be native.

Take a look at [SolveSpace](https://github.com/solvespace/solvespace) CAD for a great example of this approach.

**Comment by cody271** [comment]
> Maybe drawing custom controls would work if you could find a way to render the title bar natively?
> Idk if those methods are exposed on mac or not.

Yes emulating the native platform theme in self drawn controls _is_ possible .. Qt can do this on Mac, for example.

I don't recommend this approach. Not only is it extremely complex, also uses undocumented and unsupported APIs which are subject to change at any time. Probably not worth it.

**Comment by jwells89** [comment]
Because MDI more or less emulates the application-centric windowing model of macOS on desktops with a window-centric paradigm, my suggestion for behavior on macOS would be to do nothing except use the menubar assigned to the MDI window for all windows without a menubar explicitly assigned, since macOS already implements the remaining MDI behaviors even if it looks a bit different (no larger enclosing window).

On nixlike desktops it's a bit more complicated since conventions differ between major DEs (MDI would feel rather strange under GNOME for example). Perhaps the behavior could be a toggle that the developer could choose to surface to users.

**Comment by cody271** [comment]
> On nixlike desktops it's a bit more complicated since conventions differ between major DEs (MDI might feel rather strange under GNOME for example). Perhaps the behavior could be a toggle that the developer could choose to surface to users.

Tab based MDI emulation like WxWidgets does is the most portable fallback for Unix, but a toggle for more proper MDI implementation where DE supported is certainly reasonable.

**Comment by Shadowblitz16** [comment]
I am actually wondering if this should just be a full blown docking suggestion.

The issue is getting it to look good on all supported platforms.

I think windows could just be panels with a decently designed custom drawn title bar, same with tabs and docking tips could just be made from tooltips.

I am on vacation right now, when I get home I can draw up some examples on how they would look.

If I forgot ping me here.
